semgrep-core/src/engine/Selector.ml

```diff
@@ -0,0 +1,130 @@
+module R = Rule
+module AR = Augmented_range
+module G = AST_generic
+module M = Metavariable
+module PM = Pattern_match
+module RP = Report
+
+type selector = {
+  mvar : M.mvar;
+  pid : int;
+  pstr : string;
+  lazy_matches : RP.times RP.match_result Lazy.t;
+}
+
+type sformula =
```

Comment:
I would name the module and the type a bit different, something derived from "optimized formula". Looking into the future, I imagine that we will analyze the formula and produce a new "optimized" one. So this would not be specific to the `pattern: $X` case.

Reply:
yep, I agree.

Reply:
Actually the best would be to move this Optimize_formula.ml or Specialize_formula.ml in src/optimizing/
to keep the engine/ clean.